cmd/crowdsec-cli/setup.go

```diff
@@ -0,0 +1,312 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+	kyaml "sigs.k8s.io/yaml"
+
+	"github.com/crowdsecurity/crowdsec/pkg/csconfig"
+	"github.com/crowdsecurity/crowdsec/pkg/setup"
+)
+
+// NewSetupCmd defines the "cscli setup" command.
+func NewSetupCmd() *cobra.Command {
+	cmdSetup := &cobra.Command{
+		Use:               "setup",
+		Short:             "Tools to configure crowdsec",
+		Long:              "Manage hub configuration and service detection",
+		Args:              cobra.MinimumNArgs(0),
+		DisableAutoGenTag: true,
+	}
+
+	//
+	// cscli setup detect
+	//
+	{
+		cmdSetupDetect := &cobra.Command{
+			Use:               "detect",
+			Short:             "detect running services, generate a setup file",
+			DisableAutoGenTag: true,
+			RunE:              runSetupDetect,
+		}
+
+		defaultServiceDetect := csconfig.DefaultConfigPath("hub", "detect.yaml")
+
+		flags := cmdSetupDetect.Flags()
+		flags.String("detect-config", defaultServiceDetect, "path to service detection configuration")
+		flags.Bool("list-supported-services", false, "do not detect; only print supported services")
+		flags.StringSlice("force-unit", nil, "force detection of a systemd unit (can be repeated)")
+		flags.StringSlice("force-process", nil, "force detection of a running process (can be repeated)")
+		flags.StringSlice("skip-service", nil, "ignore a service, don't recommend hub/datasources (can be repeated)")
+		flags.String("force-os-family", "", "override OS.Family: one of linux, freebsd, windows or darwin")
+		flags.String("force-os-id", "", "override OS.ID=[debian | ubuntu | , redhat...]")
+		flags.String("force-os-version", "", "override OS.RawVersion (of OS or Linux distribution)")
+		flags.Bool("snub-systemd", false, "don't use systemd, even if available")
+		flags.Bool("yaml", false, "output yaml, not json")
```

Comment:
Any reason we would default to json rather than yaml?

Reply:
Answered my own question once I saw bats-tests

Reply:
I don't have strong opinions, but here's the rationale. JSON is easier to copy/paste (no corruption due to indentation) and programmatically manipulate or validate in case of IaC automation. In the other cases, it's an intermediary file that does not get stored anywhere, it doesn't need to be readable or to have comments. If it does (as in wizard.sh) there's the yaml option. As a rule of thumb, for machine-to-machine communication: json; machine-human-machine: yaml